internal/hcp/internal/controllers/telemetrystate/controller.go

```diff
@@ -0,0 +1,168 @@
+// SPDX-License-Identifier: BUSL-1.1
+
+package telemetrystate
+
+import (
+	"context"
+
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	"google.golang.org/protobuf/types/known/anypb"
+
+	"github.com/hashicorp/consul/internal/controller"
+	"github.com/hashicorp/consul/internal/controller/dependency"
+	"github.com/hashicorp/consul/internal/hcp/internal/controllers/link"
+	"github.com/hashicorp/consul/internal/hcp/internal/types"
+	"github.com/hashicorp/consul/internal/resource"
+	pbhcp "github.com/hashicorp/consul/proto-public/pbhcp/v2"
+	"github.com/hashicorp/consul/proto-public/pbresource"
+)
+
+var (
+	globalID = &pbresource.ID{
+		Name:    "global",
+		Type:    pbhcp.TelemetryStateType,
+		Tenancy: &pbresource.Tenancy{},
+	}
+)
+
+const MetaKeyDebugSkipDeletion = StatusKey + "/debug/skip-deletion"
+
+func TelemetryStateController(hcpClientFn link.HCPClientFn) *controller.Controller {
+	return controller.NewController(StatusKey, pbhcp.TelemetryStateType).
+		WithWatch(pbhcp.LinkType, dependency.ReplaceType(pbhcp.TelemetryStateType)).
+		WithReconciler(&telemetryStateReconciler{
+			hcpClientFn: hcpClientFn,
+		})
+}
+
+type telemetryStateReconciler struct {
+	hcpClientFn link.HCPClientFn
+}
+
+func (r *telemetryStateReconciler) Reconcile(ctx context.Context, rt controller.Runtime, req controller.Request) error {
+	// The runtime is passed by value so replacing it here for the remainder of this
+	// reconciliation request processing will not affect future invocations.
+	rt.Logger = rt.Logger.With("resource-id", req.ID, "controller", StatusKey)
+
+	rt.Logger.Trace("reconciling telemetry-state")
+
+	// First get the link resource in order to build a hcp client. If the link resource
+	// doesn't exist then the telemetry-state should not exist either.
+	res, err := getLinkResource(ctx, rt)
+	if err != nil {
+		rt.Logger.Error("failed to lookup Link resource", "error", err)
+		return err
+	}
+	if res == nil {
+		return r.ensureStateDeleted(ctx, rt)
+	}
+
+	// Check that the link resource indicates the cluster is linked
+	// If the cluster is not linked, the telemetry-state resource should not exist
+	if linked, reason := link.IsLinked(res.GetResource()); !linked {
+		rt.Logger.Trace("cluster is not linked", "reason", reason)
+		return r.ensureStateDeleted(ctx, rt)
+	}
+
+	hcpClient, err := r.hcpClientFn(res.Data)
+	if err != nil {
+		rt.Logger.Error("error creating HCP Client", "error", err)
+		return err
+	}
+
+	// Get the telemetry configuration and observability scoped credentials from hcp
+	tCfg, err := hcpClient.FetchTelemetryConfig(ctx)
+	if err != nil {
+		rt.Logger.Error("error requesting telemetry config", "error", err)
+		return err
+	}
+	clientID, clientSecret, err := hcpClient.GetObservabilitySecret(ctx)
+	if err != nil {
+		rt.Logger.Error("error requesting telemetry credentials", "error", err)
+		return nil
+	}
+
+	// TODO allow hcp client config override from hcp TelemetryConfig
+	hcpCfg := res.GetData().HcpConfig
```

Comment:
Is it an intentional choice to directly access the `HcpConfig` field instead of using the `.GetHcpConfig()` function? I just ask because it seems safer to me to use the function because there's no chance of panic if for some reason `res.GetData()` is `nil`.

Reply:
relevant linter for this: https://github.com/ghostiam/protogetter